devito/ir/clusters/algorithms.py

```diff
@@ -471,7 +471,8 @@ def normalize_reductions(cluster, sregistry, options):
             processed.extend([e.func(v, e.rhs, operation=None),
                               e.func(e.lhs, v)])
 
-        elif e.is_Reduction and e.lhs.is_Symbol and opt_mapify_reduce:
+        elif e.is_Reduction and e.lhs.is_Symbol and opt_mapify_reduce \
+                and not cluster.is_sparse:
```

Comment:
this pass is getting tricky...

maybe I would split it into two passes, `normalize_reductions_sparse` and `normalize_reductions_dense`. 

If u agree, to do that, change `@cluster_pass(mode='all')` to `@cluster_pass(mode='sparse')` and `@cluster_pass(mode='dense')` respectively

Since you're doing it... note that instead of

```
 dims = [d for d, v in cluster.properties.items() if PARALLEL_IF_ATOMIC in v]
```

you could add `is_parallel_atomic`, akin to https://github.com/devitocodes/devito/blob/master/devito/ir/support/properties.py#L208


Anyway, bottom line is: by decoupling this pass into two pass, things get way simpler

Reply:
done